lib/writer.js

```diff
@@ -13,16 +13,19 @@
       return it.value;
     };
     function type(it) {
-      if (typeof it === "string") return opt.type(it); // XXX should maintain some context
       let ret = extended_attributes(it.extAttrs);
-      if (it.union) ret += `(${it.idlType.map(type).join(" or ")})`;
+      if (it.prefix) ret += it.prefix.trivia + it.prefix.value;
+      if (it.union) ret += `${it.trivia.open}(${it.idlType.map(type).join("")}${it.trivia.close})`;
       else {
-        if (it.generic) ret += `${it.generic}<`;
-        if (Array.isArray(it.idlType)) ret += it.idlType.map(type).join(", ");
-        else ret += type(it.idlType);
-        if (it.generic) ret += ">";
+        ret += it.trivia.base;
+        if (it.generic) ret += `${it.generic.value}${it.generic.trivia.open}<`;
+        if (Array.isArray(it.idlType)) ret += it.idlType.map(type).join("");
+        else ret += typeof it.idlType === "string" ? it.baseName : type(it.idlType);
```

Comment:
Can we make generics always be arrays or either 1 or 2 length? Then we don't need the type check.

Reply:
Alternative, if it's not an array, cast it into an array?